provider/gateway/client.go

```diff
@@ -38,22 +44,66 @@ type ServiceLogMessage struct {
 }
 
 type ServiceLogs struct {
-	Stream <-chan ServiceLogMessage
+	Stream  <-chan ServiceLogMessage
+	OnClose <-chan string
 }
 
+type ClientOptions func(Client)
+
 // NewClient returns a new Client
-func NewClient() Client {
-	return &client{
-		hclient: http.DefaultClient,
+func NewClient(cctx cclient.Context, addr sdk.Address, certs []tls.Certificate) (Client, error) {
+	pclient := pmodule.AppModuleBasic{}.GetQueryClient(cctx)
+
+	res, err := pclient.Provider(context.Background(), &ptypes.QueryProviderRequest{Owner: addr.String()})
+	if err != nil {
+		return nil, err
 	}
+
+	uri, err := url.Parse(res.Provider.HostURI)
+	if err != nil {
+		return nil, err
+	}
+
+	cl := &client{
+		host:    res.Provider.HostURI,
+		addr:    addr,
+		qclient: cmodule.AppModuleBasic{}.GetQueryClient(cctx),
+	}
+
+	tlsConfig := &tls.Config{
+		ServerName:            uri.Host,
+		Certificates:          certs,
+		InsecureSkipVerify:    true,
```

Comment:
Shouldn't this be false? Otherwise we can get MITM attack'd

Reply:
No, `InsecureSkipVerify` must be set `true` as root of trust is build in `VerifyPeerCertificate` and it won't be invoked when `InsecureSkipVerify: false` as TLS library will go through normal certificate validation process and fail.
Basically we do normal certificate verification in `VerifyPeerCertificate`